react/react_testing/mocking_callbacks_and_components.md

```diff
@@ -0,0 +1,177 @@
+### Introduction
+
+We've already covered some basics of React testing. Now's the time to dive deeper.
+
+In this lesson, we'll learn about mocking. Furthermore, we'll discuss a React component from this ([theodinproject.com](https://theodinproject.com)) website and understand how React tests are written in a real world application.
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+* Carry out mocks in the context of React testing.
+
+### What is Mocking?
+
+If you've been following along with our lessons so far, the concept of mocking has already been introduced in an earlier section and you might have even incorporated mocks in your [Battleship project in the Testing JavaScript section of this course](https://www.theodinproject.com/lessons/javascript-battleship). Let's look at how mocks will help in testing React components.
+
+#### Testing Callback Handlers
+
+Callbacks are ubiquitous. Every avenue of user interaction involves callbacks. Sometimes they're passed in as props to alter state of the parent component. Consider this simple input component:
+
+~~~jsx
+// FavoriteInput.jsx
+
+import React from "react";
+
+const FavoriteInput = ({ onChange: onInputChange, id }) => {
+  const inputHandler = (event) => onInputChange(event.target.value);
+
+  return (
+    <label htmlFor={id}>
+      What is your favorite wild animal?
+      <input id={id} onChange={inputHandler} />
```

Comment:
Small nit as there's nothing inherently wrong with this as an example for what we're trying to demonstrate, but it seems a bit odd to me that this input is only half controlled since it doesn't have a value prop.

Reply:
We could the `value` prop, though now we'll have to slightly change the tests for it. Additionally, we'll also have to write another test case. Do you think it's worth it?

Reply:
I go back and forth about it. A couple alternative resolutions:

1. Change the input to a button or something that's simpler than an input anyways, this would probably be my preferred approach since I think it would really better equip us to focus in on what we're really trying to teach here.
2. Add a note that this isn't a typical real world implementation just so that people are aware of that. Also I think we should probably remove the final test in this case, as currently we're asserting against the value which is just being managed by the input itself.

Reply:
I agree with @wise-king-sullyman on this. It would be better to just use a button, since with this we're getting into controlled vs. uncontrolled components which might be out of scope for this basic example.